homeassistant/components/motionblinds_ble/cover.py

```diff
@@ -0,0 +1,283 @@
+from homeassistant.helpers.device_registry import CONNECTION_BLUETOOTH, DeviceInfo
+from homeassistant.helpers.entity_platform import AddEntitiesCallback
+
+from .const import (
+    ATTR_CONNECTION,
+    CONF_BLIND_TYPE,
+    CONF_MAC_CODE,
+    DOMAIN,
+    ICON_VERTICAL_BLIND,
+    MANUFACTURER,
+)
+
+_LOGGER = logging.getLogger(__name__)
+
+
+@dataclass(frozen=True)
+class MotionCoverEntityDescription(CoverEntityDescription):
+    """Entity description of a cover entity with default values."""
+
+    key: str = field(default=CoverDeviceClass.BLIND.value, init=False)
+    translation_key: str = field(default=CoverDeviceClass.BLIND.value, init=False)
+    device_class: CoverDeviceClass = field(default=CoverDeviceClass.SHADE, init=True)
+
+
+COVER_TYPES: dict[str, MotionCoverEntityDescription] = {
+    MotionBlindType.ROLLER.name: MotionCoverEntityDescription(),
+    MotionBlindType.HONEYCOMB.name: MotionCoverEntityDescription(),
+    MotionBlindType.ROMAN.name: MotionCoverEntityDescription(),
+    MotionBlindType.VENETIAN.name: MotionCoverEntityDescription(
+        device_class=CoverDeviceClass.BLIND
+    ),
+    MotionBlindType.VENETIAN_TILT_ONLY.name: MotionCoverEntityDescription(
+        device_class=CoverDeviceClass.BLIND
+    ),
+    MotionBlindType.DOUBLE_ROLLER.name: MotionCoverEntityDescription(),
+    MotionBlindType.CURTAIN.name: MotionCoverEntityDescription(
+        device_class=CoverDeviceClass.CURTAIN
+    ),
+    MotionBlindType.VERTICAL.name: MotionCoverEntityDescription(
+        device_class=CoverDeviceClass.CURTAIN, icon=ICON_VERTICAL_BLIND
+    ),
+}
+
+
+async def async_setup_entry(
+    hass: HomeAssistant, entry: ConfigEntry, async_add_entities: AddEntitiesCallback
+) -> None:
+    """Set up blind based on a config entry."""
+
+    blind_type = BLIND_TO_ENTITY_TYPE[entry.data[CONF_BLIND_TYPE]]
+    device = hass.data[DOMAIN][entry.entry_id]
+    blind = blind_type(device, entry)
+
+    async_add_entities([blind])
+
+
+class GenericBlind(CoverEntity):
+    """Representation of a blind."""
+
+    _attr_is_closed: bool | None = None
+    _attr_should_poll = False
+
+    _device: MotionDevice
+
+    def __init__(self, device: MotionDevice, entry: ConfigEntry) -> None:
+        """Initialize the blind."""
+        _LOGGER.debug(
+            "(%s) Setting up %s cover entity (%s)",
+            entry.data[CONF_MAC_CODE],
+            entry.data[CONF_BLIND_TYPE],
+            BLIND_TO_ENTITY_TYPE[entry.data[CONF_BLIND_TYPE]].__name__,
+        )
+        super().__init__()
+        self._device = device
+        self._device.register_running_callback(self.async_update_running)
+        self._device.register_position_callback(self.async_update_position)
+        self._device.register_connection_callback(self.async_update_connection)
+        self.entity_description = COVER_TYPES[entry.data[CONF_BLIND_TYPE]]
+        self.config_entry: ConfigEntry = entry
+
+        self._attr_name: str = device.display_name
```

Comment:
Does that have the same effect? The reason I set the name is to give each blind (device) a unique name such as `Motionblind FFFF` where FFFF is the mac code

Reply:
Yes

Reply:
For the upcoming platforms should I then also set the name using `DeviceInfo`? If so, doesn't that name all of the entities to `Motionblind FFFF`? I want only the device and the cover entity to have the name `Motionblind FFFF` and the others to get their name from strings.json `_attr_has_entity_name` and thus not from the `DeviceInfo`
```
DeviceInfo(
            ...
            name=device.display_name,
)```

Reply:
All entities will start there name with the device name and what you set in _attr_name, but perferably using a translation key will be appended to the device name.
So if the device name is "Livingroom blind" and you add an entity with _attr_name = "Battery", it will show up in Home Assistant as "Livingroom blind Battery".

See https://developers.home-assistant.io/docs/core/entity/#entity-naming

And you should set the has_entity_name attribute to True

Reply:
Is it mandatory to give all other entities the `name` in `DeviceInfo` as is done for the cover entity? Personally I don't think it adds any value to have all entities named `Motionblind FFFF Battery`, `Motionblind FFFF Speed`, `Motionblind FFFF Calibrated` etc. 

And if I understand the docs, setting has_entity_name to True and not providing a key in the translations should just result in the device name `Motionblind FFFF` right?

Reply:
I think it is indeed mandatory, because if you setup 2 blinds how would you otherwise keep the two battery sensors appart?

Note that HomeAssistant will not show the Device name in front if not needed, so for instance on the Device page it will just show "Battery" since the user will already know it is viewing "Motionblind FFFF".

> And if I understand the docs, setting has_entity_name to True and not providing a key in the translations should just result in the device name Motionblind FFFF right?

It will, but it will also result in an error message telling you you need to explicitly set _attr_name = None.

Reply:
> Note that HomeAssistant will not show the Device name in front if not needed, so for instance on the Device page it will just show "Battery" since the user will already know it is viewing "Motionblind FFFF".

Ah good to know, then I agree setting the `DeviceInfo` name for all entities to `Motionblind FFFF` is fine.

So in this case I should set _attr_has_entity_name to True and _attr_name to None for the cover entity, and then for the other entities only set _attr_has_entity_name and then provide it in the string.json right, without setting _attr_name?

Reply:
Correct, of course you will need to set "translations_key" in the entity_description for the other entities.
And once you will be adding more entities, we will make a base entity class that sets all common properties like _attr_has_entity_name etc.

Reply:
Changed in https://github.com/home-assistant/core/pull/109497/commits/22f5a8c8f041b8e869080607dcc26a46eeb7c95b

Isn't it useful to already create a base class in this PR for the entities, then the next PRs can all be created simultaneously. Otherwise we'll have to wait for another PR that contains the base entity to get merged before the other PRs can be created

Reply:
Sure, that would be nice.
Make a new file entity.py and add all common stuff there.
You can take https://github.com/home-assistant/core/blob/dev/homeassistant/components/reolink/entity.py as example.
Altought that uses a dataUpdate Coordinator, not yet sure if we will need that for this integration.

The updates of this integrations are a bit out of the ordinary...

Reply:
Thanks, added it in https://github.com/home-assistant/core/pull/109497/commits/18f9a4a14ddbe2b1ec8249fbf4ff09804a438420